### PR TITLE
LLM: Improve LLM fallback

### PR DESCRIPTION
**Pull request by pradh**
Generally, if a query can likely be handled using heuristic, we do that.  Only rarely do we fallback to LLM, and when we do, we are explicit about what from LLM-based detection we actually use (only place, only variable, or both).

Concretely, based on what the heuristic approach first detects, here are the rules for LLM fallback:

1) If no SV, SIZE and EVENT classification was detected in this query, and no SV in the context too, fallback to LLM for SV.

2) If no place and COUNTRY type was detected in this query, and no place in the context too, fallback to LLM for place.

3) For a multi-SV query (that we're sufficiently confident is multi-SV), fallback when we have >2 SVs, or if there's no correlation/comparison (because [poverty vs. asthma] queries we can handle better with heuristics), fallback to LLM for SV.

Also, refactor the detection routing logic to a separate file instead of in `api.py`.

### CODE REVIEW

File: server/lib/nl/detection/types.py [hunk 1]
```diff
@@ -346,6 +346,27 @@ class NLClassifier:
   attributes: List[ClassificationAttributes]
 
 
+class ActualDetectorType(str, Enum):
+  """Enum to represent detector types"""
+  Heuristic = "Heuristic Based"
+  LLM = "LLM Based"
+  # No fallback
```
Comment (jehangiramjad) [inline server/lib/nl/detection/types.py:353]: nit: understood what this is trying to do but I wonder if there should be two enums: one which decides whether it's a Hybrid or Not and the other which has the actual detector type used (LLM, Heuristic, LLMPlace, LLMVar).

**Review by jehangiramjad — APPROVED** [review]
Thanks! Some clarification and documentation type comments.

File: server/lib/nl/detection/llm_fallback.py [hunk 2]
```diff
@@ -13,116 +13,160 @@
...
+                                   cl.type == ClassificationType.CORRELATION
+                                   for cl in d.classifications):
+    ctr.info('info_fallback_dual_sv_correlation', '')
+    return False
+
+  return True
 
-  # TODO: Consider if we should skip this altogether.
 
+# Returns 'YES' *if* sv-parts of the `mult_sv` are delimited by a place
```
[the remark this answers is not in the document]
Comment (pradh) [reply server/lib/nl/detection/llm_fallback.py]: Added examples for each case.

File: server/lib/nl/detection/llm_fallback.py [hunk 3]
```diff
@@ -13,116 +13,160 @@
...
+# in `places_mentioned`.  If the sv or place sub-string cannot be found
+# in the query, returns 'UNSURE'.
+def _is_multi_sv_delimited(d: Detection, places_mentioned: List[str],
+                           multi_sv: dvars.MultiVarCandidate,
+                           ctr: counters.Counters) -> str:
```
[the remark this answers is not in the document]
Comment (pradh) [reply server/lib/nl/detection/llm_fallback.py]: That's right, the caller doesn't need to know NO vs. UNSURE, and previously the function returned a bool... switched it back to bool.

File: server/lib/nl/detection/types.py (hunk 1 above)
Comment (pradh) [reply server/lib/nl/detection/types.py:353]: This enum is basically just a stand-in for the returned string, so maybe 1 enum is simpler overall?

**Review by pradh — COMMENTED** [review]
Thanks for the review Jehangir!

**Pull request merged**